Development 3.0: Bind stuff

cclerget opened the pull request:
**Description of the Pull Request (PR):**

Work in progress :
- Bind system
- Overlay/Underlay support
- Session management

cclerget committed:
A proto things

cclerget committed:
Working prototype with overlay and userbinds

cclerget committed:
Remove buildcfg reference and use session directory

cclerget committed:
run before overlaylower rather

cclerget committed:
Fix overlay file/directory creation

cclerget committed:
Merge remote-tracking branch 'upstream/development-3.0' into development-3.0-prototype

cclerget committed:
Add overlay/underlay layers for session

cclerget committed:
Complete passwd/group mount

cclerget committed:
Remove useless session layer method

cclerget committed:
Session tmpfs size. Remount proc/sys

cclerget committed:
More binds

cclerget committed:
Bind system. Bug fixes

cclerget committed:
Add hostfs configuration check

cclerget committed:
Resolved merge conflict

cclerget committed:
Replace old OverlayTag by LayerTag

cclerget committed:
Reordering mount tags

cclerget committed:
Fix writable image support

cclerget committed:
Add default temporary overlay upper layer

cclerget committed:
Overlay cleanup

cclerget committed:
Replace os.Mkdir by fs helper one

cclerget committed:
Move overlay upper/work handler in container.go

cclerget committed:
Finalize home bind mount

cclerget committed:
Relax remount with user namespace for testing. Fix no-home. Disable overlay and userns

cclerget committed:
Fix remount. Add image checks. Add overlay image ro option

cclerget committed:
Merge remote-tracking branch 'upstream/development-3.0' into development-3.0-prototype

cclerget committed:
Remove bugged code

cclerget committed:
Fix panic when image path doesn't exist

cclerget committed:
Fix smaster bug

cclerget committed:
Fix oversights in some mount checks

cclerget committed:
Create passwd/group files after overlay setup

cclerget committed:
Fix underlay bugs

bauerm97 commented on src/pkg/util/fs/helper.go:
```diff
@@ -63,3 +64,28 @@ func IsSuid(name string) bool {
+
+// RootDir returns the root directory of path (rootdir of /my/path is /my).
+// Returns "." if path is empty
+func RootDir(path string) string {
```
I think we should add a few sanity test cases for this function in `helper_test.go`

cclerget replied:
Agreed

bauerm97 commented on src/runtime/engines/common/oci/config.go:
```diff
@@ -30,6 +30,6 @@ func (c *Config) UnmarshalJSON(b []byte) error {
 		return err
 	}
-	c.Generator = generate.NewFromSpec(&c.Spec)
+	c.Generator = generate.Generator{Config: &c.Spec}
```
Why is this change necessary? Just curious

cclerget replied:
Eduardo has pointed that NewFromSpec is marked as deprecated https://github.com/opencontainers/runtime-tools/blob/699dbced9a52c2d60213d8f22d6e170840d69c06/generate/generate.go#L234

bauerm97 replied:
Good catch @ArangoGutierrez

bauerm97 reviewed (changes requested):
Can we make sure that Makefile.stub is only installing the necessary dir? (Since I believe now everything happens under `$(SESSIONDIR)`)

Also this isn't directly from this PR, but `src/pkg/util/fs/layout/manager.go`, can you change `func (m *Manager) GetPath(path string) (string, error)` to use `filepath.Join()`? Thanks!

This is just a first pass review looking at some code. I have yet to fully review `container.go`. And I have yet to do testing on functionality.

cclerget committed:
Remove unused installation directories

cclerget committed:
Use of filepath.Join rather than string concatenation in Manager GetPath

cclerget committed:
Add fs helper tests and fix RootDir bug

cclerget committed:
Merge remote-tracking branch 'upstream/development-3.0' into development-3.0-prototype

cclerget committed:
Fix seccomp test failure

cclerget committed:
Fix underlay directive default value in configuration template

cclerget committed:
Fix build engine by replacing CONTAINER_FINALDIR with SESSIONDIR

cclerget committed:
Fix issues reported by Michael

cclerget committed:
Reset FS uid to current uid after privileges esacalation

cclerget committed:
Remove useless /dev remount.
Remove redundant bind from underlay.
Remove user namespace check on /tmp remount.

cclerget committed:
Remove unecessary destination check in addCwdMount
